File: python/src/cellaflow/idempotency.py
import hashlib
import rfc8785
from typing import Any, Optional


from enum import Enum


class IdempotencyScope(Enum):
    SCOPE_UNSPECIFIED = 0
    SCOPE_SESSION_WIDE = 1  # Shared across all agents in the session (Default)
    SCOPE_AGENT_PRIVATE = 2  # Isolated to the executing agent
    SCOPE_STEP_LOCAL = 3  # Isolated to the specific superstep / node
    # Shared across *sessions* within a declared coordination domain.
    # The only scope whose key omits session_id, so agents running different
    # workflows in different sessions can deduplicate one shared side effect.
    SCOPE_SHARED = 4


def _hash_inputs(*args: Any, **kwargs: Any) -> str:
    """
    Hashes the inputs using RFC 8785 Canonical JSON and SHA-256.
    Returns the first 16 bytes of the hash, hex-encoded (32 characters).
    """
    payload = {"args": list(args), "kwargs": kwargs}

    # Serialize to canonical JSON bytes according to RFC 8785
    canonical_bytes = rfc8785.dumps(payload)  # type: ignore[arg-type]

    # Hash using SHA-256
    hash_obj = hashlib.sha256(canonical_bytes)

    # Return the first 16 bytes hex-encoded
    return hash_obj.digest()[:16].hex()


def derive_idempotency_key(
    session_id: str,
    workflow_version: str,
    step_sequence: int,
    agent_id: str,
    tool_name: str,
    scope: IdempotencyScope,
    coordination_id: Optional[str],
    *args: Any,
    **kwargs: Any,
) -> str:
    """
    Derives the canonical idempotency key for a step or tool execution.
    Format varies based on IdempotencyScope.
    """
    inputs_hash = _hash_inputs(*args, **kwargs)

    if scope == IdempotencyScope.SCOPE_SHARED:
        # The only scope that omits session_id, so agents in different
        # sessions converge on one key. It also omits workflow_version, because
        # the whole point is that *different* workflows share the operation and
        # they will not be on the same version.
        #
        # coordination_id is what keeps this from being too wide. Without it two
        # unrelated callers of send_email(to=X) would deduplicate, suppressing
        # one of them silently. It is required, and the caller must choose it.
        if not coordination_id:
            raise ValueError(
                "SCOPE_SHARED requires a coordination_id naming the work being "
                "shared -- a ticket, task, or tenant id. Pass it when invoking "
                "the workflow, e.g. my_agent(..., _coordination_id='ticket-4417'). "
                "It has no default: a shared one would deduplicate unrelated "
                "callers that happen to make the same call."
            )
        return f"shared:{coordination_id}:{tool_name}:{inputs_hash}"

    seq_part = "session_wide"
    agent_part = "session_wide"

    if scope == IdempotencyScope.SCOPE_AGENT_PRIVATE:
        agent_part = agent_id
    elif scope == IdempotencyScope.SCOPE_STEP_LOCAL:
        seq_part = str(step_sequence)
        agent_part = agent_id

    return (
        f"{session_id}:{workflow_version}:{seq_part}:"
        f"{agent_part}:{tool_name}:{inputs_hash}"
    )

File: typescript/src/idempotency.ts
import canonicalize from "canonicalize";
import { createHash } from "node:crypto";

/**
 * How widely a derived idempotency key deduplicates.
 *
 * The numeric values match the Python SDK and the engine's enum. They are part
 * of the wire contract, not an implementation detail.
 */
export enum IdempotencyScope {
  UNSPECIFIED = 0,
  /** Shared across all agents in the session. The default. */
  SESSION_WIDE = 1,
  /** Isolated to the executing agent. */
  AGENT_PRIVATE = 2,
  /** Isolated to the specific superstep / node. */
  STEP_LOCAL = 3,
  /**
   * Shared across *sessions* within a declared coordination domain.
   *
   * The only scope whose key omits `sessionId`, so agents running different
   * workflows in different sessions deduplicate one shared side effect.
   */
  SHARED = 4,
}

/**
 * Hashes inputs with RFC 8785 Canonical JSON and SHA-256, returning the first
 * 16 bytes hex-encoded.
 *
 * Byte-identical to the Python SDK's `_hash_inputs`, which is what allows a
 * TypeScript agent and a Python agent to converge on one key under
 * {@link IdempotencyScope.SHARED}. Verified against Python reference hashes
 * covering unicode, escapes, nested objects, key reordering and exponent-form
 * numbers. Changing the canonicalisation or the digest length silently stops
 * cross-language deduplication, which fails by repeating the side effect rather
 * than by raising.
 */
export function hashInputs(args: unknown[], kwargs: Record<string, unknown>): string {
  const canon = canonicalize({ args, kwargs });
  if (canon === undefined) {
    throw new TypeError(
      "Tool arguments could not be canonicalised. Arguments contributing to an " +
        "idempotency key must be JSON-representable: no functions, symbols, " +
        "BigInt, undefined, or circular references.",
    );
  }
  return createHash("sha256").update(canon, "utf8").digest("hex").slice(0, 32);
}

export interface DeriveKeyOptions {
  sessionId: string;
  workflowVersion: string;
  stepSequence: number;
  agentId: string;
  toolName: string;
  scope: IdempotencyScope;
  coordinationId?: string;
  /** Positional arguments the tool was called with. */
  args?: unknown[];
  /** Named arguments, when the tool takes a single options object. */
  kwargs?: Record<string, unknown>;
  /**
   * Restricts the hash to the named keys of `kwargs`.
   *
   * Heterogeneous agents converge on one side effect precisely when they
   * disagree about everything else, so hashing everything they pass is the one
   * thing guaranteed to keep them apart.
   *
   * JavaScript has no named arguments, so a tool using `sharedOn` must take a
   * single options object; the names are read from it.
   */
  sharedOn?: readonly string[];
}

/** Derives the canonical idempotency key for a step or tool execution. */
export function deriveIdempotencyKey(opts: DeriveKeyOptions): string {
  const {
    sessionId,
    workflowVersion,
    stepSequence,
    agentId,
    toolName,
    scope,
    coordinationId,
    args = [],
    kwargs = {},
    sharedOn,
  } = opts;

  let inputsHash: string;
  if (sharedOn !== undefined) {
    const selected: Record<string, unknown> = {};
    for (const k of sharedOn) {
      if (Object.prototype.hasOwnProperty.call(kwargs, k)) selected[k] = kwargs[k];
    }
    inputsHash = hashInputs([], selected);
  } else {
    inputsHash = hashInputs(args, kwargs);
  }

  if (scope === IdempotencyScope.SHARED) {
    // The only scope that omits sessionId, so agents in different sessions
    // converge on one key. It also omits workflowVersion, because the whole
    // point is that *different* workflows share the operation and they will
    // not be on the same version.
    //
    // coordinationId is what keeps this from being too wide. Without it two
    // unrelated callers of sendEmail({to: X}) would deduplicate, suppressing
    // one of them silently. It is required, and the caller must choose it.
    if (!coordinationId) {
      throw new Error(
        "IdempotencyScope.SHARED requires a coordinationId naming the work being " +
          "shared: a ticket, task, or tenant id. Pass it when opening the session, " +
          'e.g. durableTools(config, { coordinationId: "ticket-4417" }, fn). It has ' +
          "no default: a shared one would deduplicate unrelated callers that happen " +
          "to make the same call.",
      );
    }
    return `shared:${coordinationId}:${toolName}:${inputsHash}`;
  }

  let seqPart = "session_wide";
  let agentPart = "session_wide";

  if (scope === IdempotencyScope.AGENT_PRIVATE) {
    agentPart = agentId;
  } else if (scope === IdempotencyScope.STEP_LOCAL) {
    seqPart = String(stepSequence);
    agentPart = agentId;
  }

  return `${sessionId}:${workflowVersion}:${seqPart}:${agentPart}:${toolName}:${inputsHash}`;
}
